[Docs] Add API server tuning guide (#5176)
skypilot-org/skypilot: docs/source/reference/api-server/api-server-tunning.rst

.. _sky-api-server-performance-best-practices:

SkyPilot API Server Performance Best Practices
==============================================

This page describes performance best practices for centralized SkyPilot API server in team deployment.

Tuning API server resources
---------------------------

> The concurrent requests that the API server can handle is proportional to the resources (CPU cores and memory) allocated to it. Requests are separated into two different queues and handled in a first-in-first-out manner:

Author: Sounds good!

        .. dropdown:: Why set the same value for the limits and requests?

            We recommend setting the resources limits to the same value as the requests for the following reasons:
            
            * API server depends on the resources limit to detect the available resources and decide the maximum concurrency. Setting limits larger than the requests or omitting the limits will cause the API server make aggressive concurrency decisions and may cause high resource contention on the Kubernetes node.
            * Setting the same value for the limits and requests ensures the QoS class of the API server pod being set to ``Guaranteed`` and reduce the chance of the pod being killed by the Kubernetes node when the node is under resource pressure.

            In short, setting the same value for the limits and requests sacrifices the resources utilization for stability and predictability. Pivoting to other trade-off is also possible, but we recommend to keep the memory request and limit the same in production environment to avoid potential eviction caused by Node memory pressure.

    .. tab-item:: VM Deployment

>         For VM deployment, in-place vertically scaling the API server is not supported and the API server need to be terminated and recreated to apply the new resources. This means the current state of the API server will be lost. We recommend to create an new API server instance with the new resources and gradually migrate the workload to the new API server.

Author: Do we officially supporting migrating the server with copying the `~/.sky` folder? Kubernetes Pod have `~/.sky` and `~/.ssh` persisted and cloud credentials auto mounted. Also, the environment (user, env vars, os) are consistent  between Pod recreations. I don't know whether there are corner cases for VM deployment without these guarantees. Maybe it is better to have a formal migration strategy for VM deployment, which can be a follow-up issue. Wdyt?

>         For VM deployment, in-place vertically scaling the API server is not supported and the API server need to be terminated and recreated to apply the new resources. This means the current state of the API server will be lost. We recommend to create an new API server instance with the new resources and gradually migrate the workload to the new API server.

Author: Discussed offline, migrating `~/.sky` should work. We can improve the doc about VM deployment in the future.

        Refer to :ref:`sky-api-server-cloud-deploy` for how to deploy the new API server and modify the cluster configuration before running ``sky launch``:

        .. code-block:: diff

            resources:
            -   cpus: 8+
            -   memory: 16+
            +   cpus: 16+
            +   memory: 32+

The following table shows the maximum concurrency for different resource configurations:

.. list-table::
   :widths: 1 1 2 2
   :header-rows: 1

   * - CPU
     - Memory
     - Long requests
     - Short requests
   * - 4
     - 8Gi
     - 8 Max concurrency
     - 11 Max concurrency
   * - 16
     - 32Gi
     - 32 Max concurrency
     - 60 Max concurrency
   * - 32
     - 64Gi
     - 64 Max concurrency
     - 145 Max concurrency
   * - 64
     - 128Gi
     - 128 Max concurrency
     - 299 Max concurrency
   * - 128
     - 256Gi
     - 256 Max concurrency

>      - 589 Max concurrency

Author: Like `1~10`, `1~20`, `1~50`? We can have very rough number with guts feeling I think

> 

Author: How about make this doc focus on remote API server and add this in FAQ if users encountered this?

BTW, this reminds me that the OOM behavior is actually same as 0.8.0: some of the processes are killed (or even worse, the local machine might be hung if the OOM killer acts after the memory is occupied) when launches too many concurrent requests. We can also make the UX same by:

1. decrease the `oom_score_adj` of unvicorn worker, ensure the executor get killed first when OOM happens;
2. return OOM or interrupted failure to client when executor get killed;

Queuing requests and polling status asynchronously
--------------------------------------------------

> There is no limit on the number of queued requests. So in addition to increasing the allocated resources to improve the maximum concurrency, you can also submit requests with ``--async`` flag and poll the status asynchronously to avoid blocking. For example:

Author: This is verified in https://github.com/skypilot-org/skypilot/pull/5175

.. code-block:: bash

    # Submit 2000 jobs at once without blocking
    for i in `seq 1 2000`; do

>         sky jobs launch --async job.yaml

Reviewer: ```suggestion
        sky jobs launch -y --async job.yaml
```

Author: Good catch!

    done
    # Poll the status of the jobs

>     sky jobs queue

Reviewer: ```suggestion
    watch -n 5 "sky jobs queue"
```

> The requests will be queued on the server and be processed in submission order. If you find the status is not updated for a while, you can inspect the status of the submitted requests with:

Reviewer: ```suggestion
The requests will be queued on the API server and be processed in submission order. If you find the status is not updated for a while, you can inspect the status of the submitted requests with:
```

.. code-block:: console

    $ sky api status

>     ID                                    User             Name             Created         Status

Reviewer: Can we reduce the spaces between `User` and `Name` to reduce the width of the table.

Author: Sounds good

    d9c95c7e-d248-4a7f-b72e-636511405357  alice            sky.jobs.launch  a few secs ago  PENDING
    767182fd-0202-4ae5-b2d7-ddfabea5c821  alice            sky.jobs.launch  a few secs ago  PENDING
    5667cff2-e953-4b80-9e5f-546cea83dc59  alice            sky.jobs.launch  a few secs ago  RUNNING

> 

Reviewer: Add a title: `Check logs for a request`

There should be some ``RUNNING`` requests that occupy the concurrency limit. Usually the ``RUNNING`` requests make progress and finally your requests will be processed, but if the ``RUNNING`` requests are stuck, you can inspect the request log with:

.. code-block:: console

    # Replace <request_id> with the actual request id from the ID column
    $ sky api logs <request_id>

If the request is stuck according to the log, e.g. retrying to launch VMs that is out of stock, you can cancel the request with:

> 

Reviewer: Add a title: `Cancel a request`

.. code-block:: bash

    sky api cancel <requst_id>

> Avoid concurrent logs requests

Author: We can remove this chapter after log optimization completed

------------------------------

> Though a job can run for any length of time, concurrent jobs does not occupy the concurrency. Because once a job is submitted to the cluster, it will be detached and no longer takes any resources off the API server.

Reviewer: Should we emphasize this earlier in this page to assure people that the `long requests` does not count the running jobs.

.. _sky-api-server-performance-best-practices:

SkyPilot API Server Performance Best Practices
==============================================

This page describes performance best practices for remote SkyPilot API server in :ref:`team deployment <sky-api-server>`.

Tuning API server resources
---------------------------

The number of requests that the API server can handle concurrently is proportional to the resources (CPU cores and memory) allocated to it. Internally, requests are categorized into two different types and handled in a first-in-first-out manner:

> * ``Long-running request``: request that takes long time and more resources to run, including ``launch``, ``exec``, ``jobs.launch``, etc.
> * ``Short-running request``: request that takes short time or less resources to run, including ``status``, ``logs``, etc.

Reviewer 2: ```suggestion
* ``Long-running requests``: requests that take longer time and more resources to run, including ``launch``, ``exec``, ``jobs.launch``, etc.
* ``Short-running requests``: requests that take shorter time or less resources to run, including ``status``, ``logs``, etc.
```

.. note::

>     Though a task (or job) can run for any length of time, concurrent tasks does not occupy the concurrency. Because once a task is submitted to the cluster, it will be detached and no longer takes any resources off the API server.

Reviewer 2: I think we should rephrase this somewhat, and pull it out of note box. How about something like:

Requests are queued and processed by the API server. Therefore, they only take resources off the API server when they are in queue or being processed. Once requests are processed and remote clusters start doing real work, they no longer require API server's resources or count against its concurrency limit.

For example, long-running requests for ``launch`` and ``exec`` no longer take resources off the API server once a cluster has been provisioned, or once a job has been submitted to a cluster, respectively.

Reviewer 2: Please check accuracy.

Author: Exactly! Thanks for the refinement @concretevitamin !

Reviewer 2: Sounds good, please go ahead.

It is recommended to tune the resources allocated to the API server based on the expected concurrency to avoid queuing:

.. tab-set::

    .. tab-item:: Helm Deployment

        .. code-block:: bash

            # NAMESPACE and RELEASE_NAME should be the same as the deployment step
            helm upgrade -n ${NAMESPACE} ${RELEASE_NAME} skypilot/skypilot-nightly \
                --reuse-values \
                --set apiService.resources.requests.cpu=8 \
                --set apiService.resources.requests.memory=16Gi \
                --set apiService.resources.limits.cpu=8 \
                --set apiService.resources.limits.memory=16Gi
            
        .. note:: 

>             If you specify a resources that is lower than the minimum recommended resources (4 CPUs with 8GB of memory) for team usage, an error will be raised on ``helm upgrade``. You can specify ``--set apiService.skipResourcesCheck=true`` to skip the check if performance and stability is not an issue for you scenario.

Reviewer 2: It's a bit confusing that " (4 CPUs with 8GB of memory)" is mentioned as our rec settings, but snippet above doesn't reflect this?

We should mention the rec setting in a 1-sentence paragraph.

Reviewer 2: Sounds good, please go ahead.

        Refer to :ref:`sky-api-server-cloud-deploy` for how to deploy the new API server and modify the cluster configuration before running ``sky launch``:

        .. code-block:: diff

            resources:
            -   cpus: 8+
            -   memory: 16+
            +   cpus: 16+
            +   memory: 32+

The following table shows the maximum concurrency for different resource configurations:

.. list-table::
   :widths: 1 1 3 2
   :header-rows: 1

   * - CPU
     - Memory
     - Max concurrency
     - Suggested team size
   * - 4
     - 8Gi
     - 8 Long requests, 11 Short requests
     - 1~5 users
   * - 16
     - 32Gi
     - 32 Long requests, 60 Short requests
     - 1~15 users
   * - 32
     - 64Gi
     - 64 Long requests, 145 Short requests
     - 1~30 users
   * - 64
     - 128Gi
     - 128 Long requests, 299 Short requests
     - 10~50 users
   * - 128
     - 256Gi
     - 256 Long requests, 589 Short requests
     - 10~100 users

Queuing requests and polling status asynchronously
--------------------------------------------------

> There is no limit on the number of queued requests, i.e. despite increasing the allocated resources to improve the maximum concurrency, you can also submit requests with :ref:`async<async>` (``--async``) and poll the status asynchronously to avoid blocking. For example:

Reviewer 2: ```suggestion
There is no limit on the number of queued requests. To avoid request blocking, you can either (1) allocate more resources to increase the maximum concurrency (described above), or (2) :ref:`submit requests asynchronously <async>` (``--async``) and poll the status asynchronously.

For example:
```
Do we mean this?

I still find this revised section & the sec title confusing. What do we really want to say in this section? Lmk and I can try to rephrase.

It feels like "Use asynchronous requests as much as possible"?

Author: Yes, exactly. The background is that the request is synchronized by default to keep the behavior compatible with existing users from `0.8.0`. And we do want to recommend users to "Use asynchronous requests as much as possible", but I think it is worth mention the reason in the section title so how about "Use asynchronous requests to avoid blocking"?

.. code-block:: bash

    # Submit 2000 jobs at once without blocking
    for i in `seq 1 2000`; do
        sky jobs launch -y --async job.yaml
    done
    # Poll the status of the jobs
    watch -n 5 "sky jobs queue"

The requests will be queued on the API server and be processed in submission order. If you find the status is not updated for a while, you can inspect the status of the submitted requests with:

.. code-block:: console

    $ sky api status
    ID                                    User  Name             Created         Status
    d9c95c7e-d248-4a7f-b72e-636511405357  alice sky.jobs.launch  a few secs ago  PENDING
    767182fd-0202-4ae5-b2d7-ddfabea5c821  alice sky.jobs.launch  a few secs ago  PENDING
    5667cff2-e953-4b80-9e5f-546cea83dc59  alice sky.jobs.launch  a few secs ago  RUNNING

> Check logs for a request

Reviewer 2: ```suggestion
Checking the logs of a request
```

^^^^^^^^^^^^^^^^^^^^^^^^

There should be some ``RUNNING`` requests that occupy the concurrency limit. Usually the ``RUNNING`` requests make progress and finally your requests will be processed, but if the ``RUNNING`` requests are stuck, you can inspect the request log with:

.. code-block:: console

    # Replace <request_id> with the actual request id from the ID column
    $ sky api logs <request_id>

> Cancel a request

Reviewer 2: ```suggestion
Canceling a request
```

^^^^^^^^^^^^^^^^

If the request is stuck according to the log, e.g. retrying to launch VMs that is out of stock, you can cancel the request with:

.. code-block:: bash

    sky api cancel <requst_id>

Avoid concurrent logs requests
------------------------------

> If you run ``sky logs`` to tail the logs of a task, the log tailing will keep taking off the resources of the API server as long as the task being tailed is running. So concurrent log requests will occupy the concurrency and make other requests to be delayed.

Reviewer 2: ```suggestion
If you run ``sky logs`` to tail the logs of a task, the log tailing will keep taking  resources off the API server as long as the task being tailed is still running. Thus, concurrent log requests will occupy the concurrency limit and potentially delay other requests.
```

> If you run ``sky logs`` to tail the logs of a task, the log tailing will keep taking off the resources of the API server as long as the task being tailed is running. So concurrent log requests will occupy the concurrency and make other requests to be delayed.

Reviewer 2: In this section, both 'task' and 'job' are used; can we keep one?

Author: Sure, let's stick to "task" which covers both `exec` and `jobs launch` I think
